### PR TITLE
Auto-generate filterManager rules

### PR DESCRIPTION
**Pull request by urob**
This auto-generates `filterManager` rules by parsing https://github.com/LGUG2Z/komorebi-application-specific-configuration

- `/app-rules/generate_app_rules.py`  auto-generates rules and writes them to a local copy of `DefaultFilteredWindows.cs` 
- `.github/workflows/upgrade_app_rules.yml` sets up a daily cron job (and workflow_dispatch) that runs the generator script and creates a PR if anything has changed

#641

### CODE REVIEW

**Comment by urob** [comment]
I forgot to mention, the license added to the header of the generated file is from the `komorebi` repo. The community-maintained `komorebi-application-specific-configuration` repo, which the script parses, currently does not have a license. I asked in the komorebi discord, and 1/ they totally intend for other projects to use the repo, and 2/ will add a license soon. 

I think adding the komorebi license for now and updating it once the rules repo gets a license makes sense. But open to other suggestions.

**Review by dalyIsaac — COMMENTED** [review]
You're a legend, thanks!

---

> For now, the auto-generated rules are combined with the two manual rules currently present in DefaultFilteredWindows.cs. Let me know if you prefer to separate them out in another file for easier maintenance.

Good idea! In #650 I created `DefaultFilteredWindowsKomorebi.g.cs`, and moved filter-related files to `src/Whim/Filter`.

---

> I forgot to mention, the license added to the header of the generated file is from the komorebi repo. The community-maintained komorebi-application-specific-configuration repo, which the script parses, currently does not have a license. I asked in the komorebi discord, and 1/ they totally intend for other projects to use the repo, and 2/ will add a license soon.
> 
> I think adding the komorebi license for now and updating it once the rules repo gets a license makes sense. But open to other suggestions.

Thanks for asking, it's good to hear their openness. Could you add their license to `THIRD-PARTY-NOTICES.TXT`?

---

One last thing - could you add the Python `requirements.txt` to `.github/dependabot.yml`?

File: .github/workflows/upgrade_app_rules.yml [hunk 1]
```diff
@@ -0,0 +1,72 @@
+name: Upgrade app-rules
+
+on:
+  workflow_dispatch: # Allow running on-demand
+  schedule:
+    # Runs every dat at 8:00 UTC (4:00 Eastern)
+  - cron: '0 8 * * *'
```
[the remark this answers is not in the document]
Comment (urob) [reply .github/workflows/upgrade_app_rules.yml:7]: Done. FYI, according to google 19 UTC = 05 AEST, let me know if you prefer 20 UTC

File: app_rules/generate_app_rules.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+		self.app_rules = app_rules
+
+	def generate_rules(self):
+		with open(OUTFILE, 'a') as o:
+			o.write("".join(["\n", TAB, COMMENT, self.app_name, "\n"]))
+		for r in self.app_rules:
+			FloatRule(r).add_rule()
+
+
+class FloatRule:
```
[the remark this answers is not in the document]
Comment (urob) [reply app_rules/generate_app_rules.py]: Yes, `float_identifiers` are used in komorebi to define windows that are entirely ignored (IMO somewhat of a misnomer). 

I changed the python class name to `IgnoreRule` to be more clear, and added a comment further up to clarify what `float_identifiers` are.

File: .github/workflows/upgrade_app_rules.yml [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+          git config user.email github-actions@github.com
+          git add .
+          git commit -m "Automated app-rules upgrades"
+          git push -f origin ${{ github.ref_name }}:$BRANCH_NAME
+      - name: Open pull request if needed
+        if: steps.changes.outputs.count > 0
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        # Only open a PR if the branch is not attached to an existing one
+        run: |
```
[the remark this answers is not in the document]
Comment (urob) [reply .github/workflows/upgrade_app_rules.yml:65]: I added `--label "auto app rules"` to the workflow command. I will create a test run in my fork to make sure the default `GITHUB_TOKEN` has triage access.

**Comment by urob** [comment]
Tested the new workflow in my fork. LGTM:
https://github.com/urob/Whim/pull/2

File: app_rules/requirements.txt [hunk 4]
```diff
@@ -0,0 +1,7 @@
+certifi==2023.11.17
+charset-normalizer==3.3.2
+idna==3.6
+pathlib==1.0.1
+PyYAML==6.0.1
+requests==2.31.0
+urllib3==2.1.0
```
[the remark this answers is not in the document]
Comment (dalyIsaac) [reply app_rules/requirements.txt]: Oh that makes sense

**Comment by urob** [comment]
> > The python requirements are automatically handled by the ubuntu runner started from the workflow. I have set up `actions/setup-python` to cache them, so things should be pretty fast. Maybe I am misunderstanding, but are you sure you want to also pull in the python libraries into the repo outside the runner context?
> 
> Sorry I should've explained. Dependabot is an automated bot for creating PRs with updated dependency versions.

I'm still not sure I follow. Doesn't Dependabot create PRs to the main branch? Given that there should be no need to execute the python script outside the ubuntu runner, there really should be no need to add any of the dependencies to the main branch. 

Or do you mean to update the requirements.txt file, which defines what dependencies are pulled in by the ubuntu runner? I don't see much upside to this given that the script is tested with the frozen dependencies, but could be convinced.

Sorry if I am missing something here. Maybe you can explain a bit more which files in the repo the Dependabot-generated PR would update in this case.

File: .github/workflows/upgrade_app_rules.yml [hunk 5]
```diff
@@ -38,7 +40,7 @@ jobs:
           python $APP_RULES_DIR/generate_app_rules.py
       - name: Move generated rules to destination
         run: |
-          mv $APP_RULES_DIR/DefaultFilteredWindows.cs src/Whim/Router/DefaultFilteredWindows.cs
+          mv $APP_RULES_DIR/DefaultFilteredWindowsKomorebi.g.cs src/Whim/Router/DefaultFilteredWindowsKomorebi.g.cs
```
[the remark this answers is not in the document]
Comment (urob) [reply .github/workflows/upgrade_app_rules.yml]: Not sure what happened here. I looked up the paths in the repo, no idea why I had thought `Router` is the right path. Fixed in ddd0b4f

[the remark this answers is not in the document]
Comment (dalyIsaac) [reply .github/workflows/upgrade_app_rules.yml]: I moved it - my fault 😅

**Pull request merged**